ui/views/widget/desktop_aura/desktop_drag_drop_client_ozone.cc

```diff
@@ -0,0 +1,379 @@
+        !os_exchange_data_->HasString()) {
+      return mime_type;
+    }
+    unprocessed_mime_types_.erase(unprocessed_mime_types_.begin());
+  }
+  return std::string();
+}
+
+DesktopDragDropClientOzone::DesktopDragDropClientOzone(
+    aura::Window* root_window,
+    views::DesktopNativeCursorManager* cursor_manager,
+    ui::PlatformWindow* platform_window)
+    : root_window_(root_window),
+      cursor_manager_(cursor_manager),
+      platform_window_(platform_window),
+      target_window_(nullptr),
+      delegate_(nullptr),
+      data_collector_(nullptr),
+      weak_ptr_factory_(this) {}
+
+DesktopDragDropClientOzone::~DesktopDragDropClientOzone() {}
+
+int DesktopDragDropClientOzone::StartDragAndDrop(
+    const ui::OSExchangeData& data,
+    aura::Window* root_window,
+    aura::Window* source_window,
+    const gfx::Point& root_location,
+    int operation,
+    ui::DragDropTypes::DragEventSource source) {
+  DCHECK(!source_data_);
+
+  base::RunLoop run_loop(base::RunLoop::Type::kNestableTasksAllowed);
+  quit_closure_ = run_loop.QuitClosure();
+
+  const ui::OSExchangeData::Provider* provider = &data.provider();
+  source_provider_ =
+      static_cast<const ui::OSExchangeDataProviderAura*>(provider);
+
+  gfx::ImageSkia drag_image = source_provider_->GetDragImage();
+  source_data_ = std::make_unique<ui::OSExchangeData>(data.provider().Clone());
+  aura::client::CursorClient* cursor_client =
+      aura::client::GetCursorClient(root_window);
+
+  initial_cursor_ = source_window->GetHost()->last_cursor();
+  cursor_client->SetCursor(
+      cursor_manager_->GetInitializedCursor(ui::CursorType::kGrabbing));
+
+  std::vector<std::string> mime_types;
+  mime_types.push_back(kMimeTypeText);
+  mime_types.push_back(kMimeTypeTextUTF8);
+
+  platform_window_->StartDrag(mime_types, operation, drag_image);
+  run_loop.Run();
+  return false;
+}
+
+void DesktopDragDropClientOzone::DragCancel() {
+  if (!quit_closure_.is_null())
+    quit_closure_.Run();
+  ClearDragData();
+}
+
+bool DesktopDragDropClientOzone::IsDragDropInProgress() {
+  return source_provider_ != nullptr;
+}
+
+void DesktopDragDropClientOzone::AddObserver(
+    aura::client::DragDropClientObserver* observer) {
+  NOTIMPLEMENTED();
+}
+
+void DesktopDragDropClientOzone::RemoveObserver(
+    aura::client::DragDropClientObserver* observer) {
+  NOTIMPLEMENTED();
+}
+
+void DesktopDragDropClientOzone::OnWindowDestroying(aura::Window* window) {
+  DCHECK_EQ(target_window_, window);
+  target_window_->RemoveObserver(this);
+  target_window_ = nullptr;
+}
+
+void DesktopDragDropClientOzone::OnDragEnter(
+    ui::PlatformWindow* window,
+    float x,
+    float y,
+    const std::vector<std::string>& mime_types,
+    uint32_t serial,
+    int operation) {
+  DCHECK(window);
+
+  if (mime_types.empty()) {
+    LOG(ERROR) << "Received empty mime types.";
+    return;
+  }
+
+  point_.SetPoint(x, y);
+  serial_ = serial;
```

Comment:
it doesn't look like you need a serial_ at all here. it's wayland specific and it mustn't be passed to the ozone subsystem.